geoutils/raster/delayed.py

```diff
@@ -0,0 +1,777 @@
+        if distance_unit == "pixel":
+            # Multiplying the offset by the resolution might amplify floating precision issues
+            # xoff *= self.res[0]
+            # yoff *= self.res[1]
+
+            # Using the boundaries instead!
+            xoff = xoff / self.shape[1] * (self.bounds.right - self.bounds.left)
+            yoff = yoff / self.shape[0] * (self.bounds.top - self.bounds.bottom)
+
+        shifted_transform = rio.transform.Affine(dx, b, xmin + xoff, d, dy, ymax + yoff)
+
+        return self.from_dict({"transform": shifted_transform, "crs": self.crs, "shape": self.shape})
+
+
+def _get_block_ids_per_chunk(chunks: tuple[tuple[int, ...], tuple[int, ...]]) -> list[dict[str, int]]:
+    """Get location of chunks based on array shape and list of chunk sizes."""
+
+    # Get number of chunks
+    num_chunks = (len(chunks[0]), len(chunks[1]))
+
+    # Get robust list of chunk locations (using what is done in block_id of dask.array.map_blocks)
+    # https://github.com/dask/dask/blob/24493f58660cb933855ba7629848881a6e2458c1/dask/array/core.py#L908
+    from dask.utils import cached_cumsum
+
+    starts = [cached_cumsum(c, initial_zero=True) for c in chunks]
+    nb_blocks = num_chunks[0] * num_chunks[1]
+    ixi, iyi = np.unravel_index(np.arange(nb_blocks), shape=(num_chunks[0], num_chunks[1]))
+    block_ids = [
+        {
+            "num_block": i,
+            "ys": starts[0][ixi[i]],
+            "xs": starts[1][iyi[i]],
+            "ye": starts[0][ixi[i] + 1],
+            "xe": starts[1][iyi[i] + 1],
+        }
+        for i in range(nb_blocks)
+    ]
+
+    return block_ids
+
+
+class ChunkedGeoGrid:
+    """
+    Chunked georeferenced grid class.
+
+    Associates a georeferenced grid to chunks (possibly of varying sizes).
+    """
+
+    def __init__(self, grid: GeoGrid, chunks: tuple[tuple[int, ...], tuple[int, ...]]):
+
+        self._grid = grid
+        self._chunks = chunks
+
+    @property
+    def grid(self) -> GeoGrid:
+        return self._grid
+
+    @property
+    def chunks(self) -> tuple[tuple[int, ...], tuple[int, ...]]:
+        return self._chunks
+
+    def get_block_locations(self) -> list[dict[str, int]]:
+        """Get block locations in 2D: xstart, xend, ystart, yend."""
+        return _get_block_ids_per_chunk(self._chunks)
+
+    def get_blocks_as_geogrids(self) -> list[GeoGrid]:
+        """Get blocks as geogrids with updated transform/shape."""
+
+        block_ids = self.get_block_locations()
+
+        list_geogrids = []
+        for bid in block_ids:
+            # We get the block size
+            block_shape = (bid["ye"] - bid["ys"], bid["xe"] - bid["xs"])
+            # Build a temporary geogrid with the same transform as the full grid
+            geogrid_tmp = GeoGrid(transform=self.grid.transform, crs=self.grid.crs, shape=block_shape)
+            # And shift it to the right location (X is positive in index direction, Y is negative)
```

Comment:
Why not directly have GeoGrid take as arguments "xstart" and "ystart"?

Reply:
I'm not sure how that would work for the transform, a GeoGrid is independent of a chunk size, and "xstart" and "ystart" are indexes not coordinates.
Maybe a class method `GeoGrid.from_corner_xy()` where I pass `bounds.left + xs * xres` and same for y? It's the equivalent of what's done in another order, but if it is clearer it'd be better.

Reply:
I think both are equally clear to me, I added more comments to describe the new shape + shift method already coded